platform/src/share/handle.rs

```diff
@@ -0,0 +1,77 @@
+use crate::share::List;
+use core::marker::PhantomData;
+
+/// A `Handle`'s existence indicates two things:
+///
+/// 1. The specified List exists.
+/// 2. The specified List will be cleaned up (via `Drop::drop`) rather than
+///    leaked or forgotten.
+///
+/// `Handle`s are used to call system calls, and should generally be created by
+/// using `share::scope`.
+pub struct Handle<'handle, L: List> {
+    // Handle acts like a &'handle L, so set its variance accordingly. Note that
+    // the most important lifetime -- the lifetime the share can live for -- is
+    // a parameter of L, not Handle.
+    //
+    // Additionally, _list is a private member, which prevents code outside this
+    // module from constructing a Handle without calling Handle's constructors.
+    _list: PhantomData<&'handle L>,
+}
+
+// We can't #[derive(Clone, Copy)] because derive's implementations of Clone and
+// Copy have `L: Clone` and `L: Copy` constraints, respectively. We don't want
+// those constraints, so we manually implement Clone and Copy.
+impl<'handle, L: List> Clone for Handle<'handle, L> {
+    fn clone(&self) -> Self {
+        Handle { _list: PhantomData }
+    }
+}
+
+impl<'handle, L: List> Copy for Handle<'handle, L> {}
+
+impl<'handle, L: List> Handle<'handle, L> {
+    /// Constructs a new `Handle`, typing its lifetime to the provided list.
+    ///
+    /// # Safety
+    /// The calling code must guarantee that `Drop::drop` is called on `_list`'s
+    /// pointee before `_list`'s pointee becomes invalid. In other words,
+    /// `_list`'s pointee may not be forgotten or leaked.
+    pub unsafe fn new(_list: &'handle L) -> Self {
+        Handle { _list: PhantomData }
+    }
+
+    /// Converts this `Handle` to a `Handle` of a different type. Used to create
+    /// a handle to a sub-object of this handle's `L`.
+    ///
+    /// # Safety
+    /// The calling code must guarantee that an `Other` exists, and that
```

Comment:
So both other and self will need to have separate `Drop` calls? Or should the `other` not get dropped since the new self will get dropped instead?

Reply:
I added the following to the safety comment:

```rust
/// In general, the `Other` should be contained inside the `L`, so
/// `Drop::drop` is executed on the `Other` when the `L` is dropped.
```

Does that answer your question?

Reply:
It does thanks!